modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+    using Quatf = Quat<float>;
+    using std::sin;
+    using std::cos;
+    using std::tan;
+    using std::exp;
+    using std::log;
+    using std::sinh;
+    using std::cosh;
+    using std::tanh;
+    using std::sqrt;
+    using std::asinh;
+    using std::acosh;
+    using std::atanh;
+    using std::asin;
+    using std::acos;
+    using std::atan;
+//! @} core
+}
+
+//! @cond IGNORE
+///////////////////////////////////////////////////////////////////////////////////////
+//Implementation
+namespace cv{
+    template <typename T>
+    Quat<T>::Quat(const Vec<T, 4> &coeff):w(coeff[0]), x(coeff[1]), y(coeff[2]), z(coeff[3]){}
+
+    template <typename T>
+    Quat<T>::Quat(const T qw, const T qx, const T qy, const T qz):w(qw), x(qx), y(qy), z(qz){}
+
+    template <typename T>
+    Quat<T>::Quat(const T angle, const cv::Vec<T, 3> &axis)
+    {
+        T vNorm = std::sqrt(axis.dot(axis));
+        if (vNorm < CV_QUAT_EPS)
+        {
+            CV_Error(Error::StsBadArg, "this quaternion does not represent a rotation");
+        }
+        w = std::cos(angle / 2);
+        x = std::sin(angle / 2) * (axis[0] / vNorm);
+        y = std::sin(angle / 2) * (axis[1] / vNorm);
+        z = std::sin(angle / 2) * (axis[2] / vNorm);
+    }
+
+    template <typename T>
+    Quat<T>::Quat(const Mat &R)
+    {
+        if (R.rows != 3 || R.cols != 3)
+        {
+            CV_Error(Error::StsBadArg, "Cannot matrix to quaternion: rotation matrix should be a 3x3 matrix");
+        }
+        T S;
+        T trace = R.at<T>(0, 0) + R.at<T>(1, 1) + R.at<T>(2, 2);
+        if (trace > 0)
+        {
+            S = std::sqrt(trace + 1) * 2;
+            x = (R.at<T>(1, 2) - R.at<T>(2, 1)) / S;
+            y = (R.at<T>(2, 0) - R.at<T>(0, 2)) / S;
+            z = (R.at<T>(0, 1) - R.at<T>(1, 0)) / S;
+            w = -0.25 * S;
+        }
+        else if (R.at<T>(0, 0) > R.at<T>(1, 1) && R.at<T>(0, 0) > R.at<T>(2, 2))
+        {
+
+            S = std::sqrt(1.0 + R.at<T>(0, 0) - R.at<T>(1, 1) - R.at<T>(2, 2)) * 2;
+            x = -0.25 * S;
+            y = -(R.at<T>(1, 0) + R.at<T>(0, 1)) / S;
+            z = -(R.at<T>(0, 2) + R.at<T>(2, 0)) / S;
+            w = (R.at<T>(1, 2) - R.at<T>(2, 1)) / S;
+        }
+        else if (R.at<T>(1, 1) > R.at<T>(2, 2))
+        {
+            S = std::sqrt(1.0 - R.at<T>(0, 0) + R.at<T>(1, 1) - R.at<T>(2, 2)) * 2;
+            x = (R.at<T>(0, 1) + R.at<T>(1, 0)) / S;
+            y = 0.25 * S;
+            z = (R.at<T>(1, 2) + R.at<T>(2, 1)) / S;
+            w = (R.at<T>(0, 2) - R.at<T>(2, 0)) / S;
+        }
+        else
+        {
+            S = std::sqrt(1.0 - R.at<T>(0, 0) - R.at<T>(1, 1) + R.at<T>(2, 2)) * 2;
+            x = (R.at<T>(0, 2) + R.at<T>(2, 0)) / S;
+            y = (R.at<T>(1, 2) + R.at<T>(2, 1)) / S;
+            z = 0.25 * S;
+            w = -(R.at<T>(0, 1) - R.at<T>(1, 0)) / S;
+        }
+    }
+
+    template <typename T>
+    Quat<T>::Quat(const Vec<T, 3> &rvec)
```

Comment:
A quaternion contains cos and sin of half angle, here we have full angle.
Need to do one of the following:
* divide angle by 2
* add a note into description that this function accepts half angle by axis as an argument

Reply:
Referring to the paper [Rotations with Rodrigues' vector](https://www.researchgate.net/publication/231052227_Rotations_with_Rodrigues'_vector) of formula (8),  rvec is defined as:

r = tan(psi / 2) * n

where n represents normalized axis vector. 

At line 1198, in fact, the var angle is the half angle, i.e., angle = psi / 2. I  think quaternion can also be constructed by this angle.

Reply:
Yes, I forgot that peculiarity of Rodrigues vector. In that case we should add a note into description about half angle.

Reply:
OK

Reply:
Not sure to have understood everything correctly.

---

For me, by default Rodrigues rotation vector should be `angle x axis()` (vector of size 3) and not `tan(psi / 2) * n` or `tan(angle/2) x axis()`.

I believe `rvec = angle x axis()` is the convention that is used in the [calib3d](https://docs.opencv.org/4.5.0/d9/d0c/group__calib3d.html#ga61585db663d9da06b68e70cfbf6a1eac) module, and similar to the [Rodrigues' rotation formula](https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula) wiki page.

This would be more or less consistent with `Eigen::AngleAxis` type, see for instance this [conversion code](https://gitlab.com/libeigen/eigen/-/blob/4a77eda1fdaebd7f92d587dfc0158a20dc0d2625/Eigen/src/Geometry/Quaternion.h#L560-569).

---

In any case, description of the expected Rodrigues vector is needed to avoid confusion.

Reply:
Rodrigues' rotation formula just illustrates the transformation from axis angle to the rotation vector, maybe not Rodrigues vector.
 
So far, there are three ways to represent a rotation vector
 - `angle * axis()` just used in calib3d module. It is called [rotation vector ](https://en.wikipedia.org/wiki/Axis%E2%80%93angle_representation#Rotation_vector)in wiki
-  a struct of angle and axis, in  [Eigen::AngleAxis](https://gitlab.com/libeigen/eigen/-/blob/4a77eda1fdaebd7f92d587dfc0158a20dc0d2625/Eigen/src/Geometry/AngleAxis.h)
- `tan(psi / 2) * n` ,used here, called Rodrigues vector(RF vector)

if needed, we could add a method from `angle * axis()` to quaternion. Anyway,  we will add description to avoid confusion

Reply:
Rodrigues' rotation formula just illustrates the transformation from axis angle to the rotation vector, maybe not Rodrigues vector.
 
So far, there are three ways to represent a rotation vector
 - `angle * axis()` just used in calib3d module. It is called [rotation vector ](https://en.wikipedia.org/wiki/Axis%E2%80%93angle_representation#Rotation_vector)in wiki
-  a struct of angle and axis, in  [Eigen::AngleAxis](https://gitlab.com/libeigen/eigen/-/blob/4a77eda1fdaebd7f92d587dfc0158a20dc0d2625/Eigen/src/Geometry/AngleAxis.h)
- `tan(psi / 2) * n` ,used here, called Rodrigues vector(RF vector)

if needed, we could add a method from `angle * axis()` to quaternion. Anyway,  we will add description to avoid confusion

Reply:
Indeed Rodrigues' rotation formula is about rotating a vector by an axis of rotation and an angle.

---

Still, I am not convinced about using `tan(psi / 2) * n` rotation representation. It is the first time I have seen this representation. Granted, I am not an expert.

In my opinion, `rvec` variable name should mean `angle * axis()` and match the meaning of `rvec` in the `calib3d` module. If `tan(psi / 2) * n` representation should be  supported, it should be non-ambiguous for a novice user, and no confusion should be possible with `angle * axis()` representation.

---

Is there really an usage for `tan(psi / 2) * n` rotation representation? Is it really needed to support it?

I don't think Eigen supports this representation. I don't think it is described in the [Handbook of Robotics](https://www.cc.gatech.edu/~hic/4632B-07/kinematics-draft.pdf). And I would like to see a library that supports this rotation representation to be convinced that it can be useful.

Reply:
I also vote for `axis*angle` vector instead of "true" tan-based Rodrigues vector - here and everywhere in OpenCV. But, to avoid mixing the terms, this should be called "rotation vector", not a "Rodrigues vector".
First, because it fails to represent rotations close to pi/2 around any axis.
Second, because it's more conventional and it's used everywhere, from calib3d to Eigen.

Reply:
A solution proposed by @alalek: get rid of constructors like `Quat(Vec3)`, `Quat(Mat)`,` Quat(Vec4)`, make functions like createFromAxisAngle, createFromRodrigues, createFromRotMatrix,.. instead. When passing a piece of data as an argument to constructor, it's hard to understand its meaning, named functions are less confusing.
I agree with that.

Reply:
OK. We will change it to `axis*angle` form and corresponding description